### PR TITLE
Restore energy density estimator

### PR DESCRIPTION
**Pull request by jtkrogel**
## Proposed changes

Restore functioning of the energy density estimator.  With these updates, uniform and voronoi grids work in periodic boundary conditions, as confirmed by statistical tests as well as deterministic checksums (integral of each energy density sample must reproduce the scalar energy for each block).  Checks have been performed in both the single and multi-threaded cases.  Existing in-situ tests were invaluable for the restoration.

Some issues remain to be handled: 
* Adjustments need to be made to some of the statistical tests, which have mis-estimated errorbars (by a clean factor of two).  These were likely errors in test creation relating to walker count.  
* Systems with open boundary conditions have yet to be tested fully (later PR).
* Data routing issues via TraceManager remain.  An easy workaround is to switch traces on in the input file (without writing to disk).  Given relative priorities, this remnant issue is much lower priority (perhaps never in a PR).

## What type(s) of changes does this code introduce?

- Bugfix

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?

Laptop

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'

### CODE REVIEW

File: src/Particle/LongRange/StructFact.cpp [hunk 1]
```diff
@@ -24,7 +24,7 @@ namespace qmcplusplus
 {
 //Constructor - pass arguments to KLists' constructor
 StructFact::StructFact(ParticleSet& P, RealType kc)
-    : DoUpdate(false), StorePerParticle(false), SuperCellEnum(SUPERCELL_BULK)
+    : DoUpdate(false), StorePerParticle(true), SuperCellEnum(SUPERCELL_BULK)
```
Comment (jtkrogel) [inline src/Particle/LongRange/StructFact.cpp:27]: @ye-luo What do you recommend to do with this?  Enabling this storage is necessary for streaming single particle energy information as needed for energy density and EKT.

Comment (ye-luo) [reply src/Particle/LongRange/StructFact.cpp:27]: Call ParticleSet::turnOnPerParticleSK() by the feature related consumer class constructor?

Comment (jtkrogel) [reply src/Particle/LongRange/StructFact.cpp:27]: I will give it a try.

Comment (jtkrogel) [reply src/Particle/LongRange/StructFact.cpp:27]: This just results in a segfault as if `StorePerParticle` is `false`.  Same result if called at beginning of each evaluate.

Comment (jtkrogel) [reply src/Particle/LongRange/StructFact.cpp:27]: Perhaps the resize is wrong?

Comment (jtkrogel) [reply src/Particle/LongRange/StructFact.cpp:27]: In `StructFact::resize`: `eikr_r.resize(nptcl, nkpts);`

Comment (jtkrogel) [reply src/Particle/LongRange/StructFact.cpp:27]: In `StructFact::turnOnStorePerParticle`: `eikr_r.resize(nptcl, KLists.numk);`

Comment (jtkrogel) [reply src/Particle/LongRange/StructFact.cpp:27]: Apparently not the problem.  Suggestions?

Comment (PDoakORNL) [reply src/Particle/LongRange/StructFact.cpp:27]: build debug
gdb your_unit_test

Comment (jtkrogel) [reply src/Particle/LongRange/StructFact.cpp:27]: Thanks.  I know exactly what is causing the segfault.

Comment (jtkrogel) [reply src/Particle/LongRange/StructFact.cpp:27]: The underlying issue (again) relates to state machines and order of operations, specifically the order of calls to `StructFact::resize` and `StructFact::turnOnStorePerParticle`.

Comment (jtkrogel) [reply src/Particle/LongRange/StructFact.cpp]: Found the issue.  See latest commit.

**Comment by jtkrogel** [comment]
Ready now.

**Comment by jtkrogel** [comment]
NB: this should go in ahead of #2599 in case of any interactions.

**Comment by ye-luo** [comment]
Okay to test

**Comment by ye-luo** [comment]
Could you fix `docs/readthedocs.org:nexus-workflows`?

**Comment by jtkrogel** [comment]
What is the problem with it?

**Comment by ye-luo** [comment]
> Could you fix `docs/readthedocs.org:nexus-workflows`?

CI was showing failure but it seems fine now.

**Comment by PDoakORNL** [comment]
I'm not sure how there would be interactions between this and #2599 since the legacy estimators and drivers have not been altered. I could just as easily say #2599 should go in first because that work might be affected by #2601
